src/framework/mlt_slices.c

```diff
@@ -0,0 +1,229 @@
+/**
+ * \file mlt_slices.c
+ * \brief sliced threading processing helper
+ * \see mlt_slices_s
+ *
+ * Copyright (C) 2016 Meltytech, LLC
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include "mlt_slices.h"
+
+#include <stdlib.h>
+#include <unistd.h>
+#include <pthread.h>
+
+#define MAX_SLICES 32
+#define ENV_SLICES "MLT_SLICES_COUNT"
+
+typedef struct mlt_slices_s
+{
+	int f_exit;
+	int count;
+	int readys;
+	pthread_mutex_t cond_mutex;
+	pthread_cond_t cond_var_job;
+	pthread_cond_t cond_var_ready;
+	pthread_t threads[MAX_SLICES];
+	struct
+	{
+		int jobs, done;
+		mlt_slices_proc proc;
+		void* cookie;
+	} runtime;
+} *mlt_slices;
+
+
+static void* mlt_slices_worker( void* p )
+{
+	int id, idx;
+	mlt_slices ctx = (mlt_slices)p;
+
+	while ( 1 )
+	{
+		if ( ctx->f_exit )
+			pthread_exit(0);
+
+		pthread_mutex_lock( &ctx->cond_mutex );
+		id = ctx->readys;
+		ctx->readys++;
+		pthread_cond_signal( &ctx->cond_var_ready );
+		if ( !ctx->f_exit )
+			pthread_cond_wait( &ctx->cond_var_job, &ctx->cond_mutex );
+		pthread_mutex_unlock( &ctx->cond_mutex );
+
+		if ( ctx->f_exit )
+			pthread_exit(0);
+
+		pthread_mutex_lock( &ctx->cond_mutex );
+		while ( ctx->runtime.done < ctx->runtime.jobs && !ctx->f_exit )
+		{
+			idx = ctx->runtime.done;
+			ctx->runtime.done++;
+			pthread_mutex_unlock( &ctx->cond_mutex );
+			ctx->runtime.proc( id, idx, ctx->runtime.jobs, ctx->runtime.cookie );
+			pthread_mutex_lock( &ctx->cond_mutex );
+		}
+		pthread_mutex_unlock( &ctx->cond_mutex );
+	}
+
+	return NULL;
+}
+
+/** Initialize a sliced threading context
+ *
+ * \public \memberof mlt_slices_s
+ * \param threads number of threads to use for job list
+ * \param policy scheduling policy of processing threads
+ * \param priority priority value that can be used with the scheduling algorithm
+ * \return the context pointer
+ */
+
+mlt_slices mlt_slices_init( int threads, int policy, int priority )
+{
+	pthread_attr_t tattr;
+	struct sched_param param;
+	mlt_slices ctx = (mlt_slices)calloc( 1, sizeof( struct mlt_slices_s ) );
+	char *env = getenv( ENV_SLICES );
+#ifdef _WIN32
+	int cpus = GetCurrentProcessorNumber( );
+#else
+	int cpus = sysconf( _SC_NPROCESSORS_ONLN );
+#endif
```

Comment:
This does not adequately address Darwin and BSD targets. It needs to be `if defined(WIN32) ... elif defined (__linux__) ... else ....` Maybe the else should just be 1 or some macro definition.